Refactor: Standardize Search Tools

NISH1001 opened the pull request:
### Summary :memo:

This PR introduces a significant refactoring of the entire search tool module. It establishes a new `SearchTool` base class that all specific search tools (SearxNG, Serper, Semantic Scholar, Code Search) now inherit from. This change modularizes the search process, centralizing logic for parallel query execution, result aggregation via **Reciprocal Rank Fusion (RRF)**, normalization, deduplication, and adding a new, optional reranking step at the base level.

Simultaneously, this PR overhauls the `resolvers` module, renaming `ResearchArticleResolver` to `CompositeResolver` and changing its core logic from a simple waterfall to a comprehensive "run-all-and-merge" strategy to maximize metadata enrichment.

### Details

All search tools now follow the following pattern:

  - Must implement the abstract `_arun_single_query` method.
  - The base `arun` method runs multiple queries in parallel.
  - Results from all queries are normalized and deduplicated.
  - Each query's result set is *optionally* reranked (e.g., via `cross_encoder`).
  - **Reciprocal Rank Fusion (RRF)** is applied to the final, aggregated list of results.

## Major Changes

  * **Refactor `SearchTool` Base Class (`akd/tools/search/_base.py`)**

      * The main `arun()` method is now implemented in this base class. It centrally handles:
        1.  Running multiple queries in parallel using `asyncio.gather`.
        2.  Normalizing and deduplicating all retrieved results.
        3.  Calling a new `_rerank()` method for each query's individual results (if a reranker is configured).
        4.  Aggregating all results using **Reciprocal Rank Fusion (RRF)** (via new `rrf_fusion` utility).

  * **Modular `_arun_single_query` Method**

      * All specific search tools (`SearxNG`, `Serper`, `SemanticScholar`, `CodeSearch`) were refactored to inherit from the new `SearchTool`.
      * Their primary execution logic was moved from `_arun` to the new abstract method `_arun_single_query(self, query: str, max_results: int, **kwargs)`.

  * **Reranker Integration (`akd/tools/reranker.py`)**

      * A new `create_reranker` factory function simplifies creating `RerankerTool` instances (`cross_encoder` or `none`).
      * Deduplication and sorting logic was removed from the reranker tool itself and moved to the `SearchTool` base class (using new utilities).

  * **Major Resolver Refactor (`akd/tools/resolvers/`)**

      * **Rename:** `ResearchArticleResolver` has been renamed to `CompositeResolver`.
      * **New Logic:** The `CompositeResolver` logic was changed from a "stop-at-first-success" waterfall to a "run-all-and-merge" strategy. It now runs *all* sub-resolvers to gather maximum metadata.
      * **Smart URL Selection:** The `CompositeResolver` now intelligently selects the best final URL for crawling (prioritizing PDF \> OA Host \> DOI \> Original) and stores alternatives in `extra["alternative_urls"]`.
      * **Metadata Enrichment:** `CrossRefDoiResolver` and `UnpaywallResolver` now actively enrich metadata (title, authors, date) from their API responses, not just return a URL or DOI.

  * **HTTP Client Standardization**

      * Replaced `aiohttp` with `httpx` in `SearxNG`, `SemanticScholar`, `ADSResolver`, and `UnpaywallResolver` for consistency.

  * **Code Search Refactor**

      * The `CodeSearch` tool and its tests were significantly refactored to align with the new `SearchTool` base class structure.
      * `code_search` module has been moved to `akd.tools.search.code_search.py`.
      * `CombinedSearchTool` has been renamed to `CompositeSearchTool` to maintain `akd` naming semantics.
      * `CodeSearchAgentConfig` was updated, replacing `reranker_tool` with `reranker_type` (an enum) and a `reranker_config` model.

  * **CompositeSearchTool**

      * `akd.tools.search.composite.CompositeSearchTool` is added with the ability to perform fusion of results from multiple different search tools.

    <!-- end list -->

    ```python
    from akd.tools.search import CompositeSearchTool, CompositeSearchToolConfig

    tool_1 = SearxngSearchTool(...)
    tool_2 = SerperSearchTool(...)

    search_tool = CompositeSearchTool(tools=[tool_1, tool_2], debug=True)

    output = await search_tool.arun(SearchToolInputSchema(...))

    # output.results give list of SearchResultItem
    # We can get which tool was used with `result.extra["tool"]`
    ```

## Minor Changes

  * **New Utilities:** Created `akd/tools/search/utils.py` to house shared logic for `rrf_fusion`, `deduplicate_results`, and `normalize_results`.
  * **Module Renaming:** Renamed search tool modules for clarity:
      * `searxng_search.py` -\> `searxng.py`
      * `semantic_scholar_search.py` -\> `semantic_scholar.py` (file was refactored, not renamed)
  * **Agent Refactor:** `SearchAgent` (`akd/agents/search/_base.py`) no longer inherits from `SearchTool`, simplifying its design to be a *consumer* of the tool.
  * **Pipeline Optimization:** `SearchPipeline` (`akd/tools/search/pipeline.py`) now performs a sanity check to skip the `CompositeResolver` step if the search result already contains resolved metadata (like a `pdf_url`).
  * **Schema Update:** Removed the `category` field from `SearchToolOutputSchema`.
  * **Config Loading:** Fixed a bug in `akd._base.py`'s `__set_attrs_from_config` method to correctly load nested Pydantic models and computed fields from config objects.
  * **Testing Overhaul:**
      * Refactored all unit tests for `SearxNG`, `Serper`, and resolvers to align with the new class structures and `httpx` mocking.
      * Added a new E2E test suite (`test_search_tools_e2e.py`) to test the public-facing `arun()` method and validate the full RRF and reranking flow.
      * Updated tests to call the public `arun()` method instead of the private `_arun()`.

### Checks

  - [x] Tested Changes (Same tests passing as that of `develop` branch)
  - [ ] Stakeholder Approval

NISH1001 committed:
Significantly refactor search tool module to incorporate modularity and
extensiblity

NISH1001 committed:
Fix tests for searnxg to reflect new changes

NISH1001 committed:
Fix serper tests

NISH1001 committed:
Bugfix searchpipeline abstract method

NISH1001 committed:
Refactor base search tool class to make it extensibility for any
downstream search tools

NISH1001 committed:
Bugfix semantic scholar search tool with new refactor

NISH1001 committed:
Remove category field from searchtooloutputschema

- Having the category field doesn't make sense

NISH1001 committed:
Parameterize e2e tests for search tools

NISH1001 committed:
Refactor codesearch tools to conform to new structure

- Use SearchTool baes class
-

NISH1001 committed:
Refactor code search tool and fix bugs

NISH1001 committed:
Rename searh module names to just be their respective tool name

- searxng_search.py changed to `searxng.py`
- semantic_scholar_search to semantic_scholar

NISH1001 committed:
Pass more kwargs to single query in search tool base

NISH1001 committed:
Add reranker to search tool at base class level

- Add `reranker_type` and `reranker_config` field to SearchToolConfig
- If there's no reranking applied, use NoOpRerankerTool

NISH1001 committed:
Inherit search agent only from baseagent

- Solves diamond problem

NISH1001 committed:
Add composite search tool
- Add akd.tools.search.composite.CompositeSearchTool that combines
  multiple tool results and applies fusion strategy

- Also refactor codesearchcombined

NISH1001 committed:
Fix code search agent module

NISH1001 committed:
Fix imports for code search tool

NISH1001 committed:
Bugfix circular import for code search tool

NISH1001 committed:
Bugfix model dump issue in akd base class

- Instead of doing config.model_dump() -- which was causing nested model
  to be dicts -- use field-level setter
- Also refactor code agent to use new config

NISH1001 committed:
Bugfix set attr method in akd/_base.py for computed fields

muthukumaranR commented on akd/tools/search/composite.py:
```diff
@@ -0,0 +1,244 @@
+        description=(
+            "Strategy for fusing results from multiple tools. "
+            "'direct_rrf_blackbox': Direct RRF on tool outputs (treats tools as independent blackbox rankers). "
+            "'flatten_rerank_rrf': Flatten all results, rerank globally with cross-encoder, then apply RRF (better semantic ranking)."
```
How would RRF work if we have a flat list? doesn't RRF need multiple lists?

muthukumaranR replied:
I think Flaten + rerank makes sense, for a single (original) query

NISH1001 replied:
Yes. Here, the diea is: flatten -> then rerank if multiple queries and apply rrf. Promise is : multiple queries.

NISH1001 replied:
But if it's just a single list (from single query), applying rrf won't change anything after reranking....

NISH1001 committed:
Rename CombinedCodeSearchTool to CompositeCodeSearchTool

NISH1001 committed:
Fix code search composite tool name in example

NISH1001 committed:
Add utility function to deduplicate search results

- Add `akd.tools.search.utils.deduplicate_results`

NISH1001 committed:
Merge branch 'develop' of github.com:NASA-IMPACT/accelerated-discovery into feature/search-tool-refactor-rrf

NISH1001 committed:
Add url-doi resolver static method to DOIResolver

NISH1001 committed:
Improve unpaywall resolver

- Populate other fields like pdf url, original url, authors, extra
  metadata

NISH1001 committed:
Bugfix doiresolver tests

NISH1001 committed:
Improve openaccess url resolver in unpaywall

NISH1001 committed:
Improve resolvers to update all the available metadata for the search
result items

NISH1001 committed:
Rename researcharticleresolver to CompositeResolver

NISH1001 committed:
REmove normalizer

- Now the composite resolver does the job of normalization

NISH1001 committed:
Change the default email for unpaywallresolver to akd@uah.edu

NISH1001 committed:
Improve resolover for pdf

NISH1001 committed:
Improve rrf with multiple key matching

- Repalce `key` (str) with `keys` (list[str])
- Add `rrf_keys` to searchtoolconfig

NISH1001 committed:
Improve cascaded rrf and debug log

NISH1001 committed:
Add search result normalization to searech tool

NISH1001 committed:
Enable result normalization by default in base search tool

- Disable at codesearch tool

NISH1001 committed:
Update deduplication strategy

- Use mutliple keys

NISH1001 committed:
Remove redundancy code duplication throughout code search tools

NISH1001 committed:
Sanity check pre-filled resolved metadata in search pipeline and skip if
not required

NISH1001 committed:
Bugfix searxng test

NISH1001 committed:
Add value normalizers to rrf function

- Add a dictionary str -> callable mapper

NISH1001 committed:
Move imports to top-level import for controlled search agent
